File: test/test_api_header.rb
require_relative "../lib/rgfa.rb"
require "test/unit"
TestAPI ||= Module.new

class TestAPI::Header < Test::Unit::TestCase

  def test_rgfa_header
    g = RGFA.new
    assert_equal(RGFA::Line::Header, g.header.class)
    assert_equal([], g.header.tagnames)
    g << "H\txx:i:1".to_rgfa_line
    assert_equal([:xx], g.header.tagnames)
  end

  def test_rgfa_header_line_connect
    g = RGFA.new
    line = "H\txx:i:1".to_rgfa_line
    assert_raise(RGFA::RuntimeError) { line.connect(g) }
    assert_nothing_raised { g.add_line(line) }
  end

  def test_header_version_editing
    standalone = "H\txx:i:1\tVN:Z:1.0".to_rgfa_line
    assert_nothing_raised {standalone.VN = "2.0"}
    g = RGFA.new
    g << "H\txx:i:1\tVN:Z:1.0"
    assert_nothing_raised { g.header.xx = 2}
    assert_raise(RGFA::RuntimeError) {
      g.header.VN = "2.0" }
  end

  def test_error_inconsistent_definitions
    g = RGFA.new
    g << "H\txx:i:1"
    assert_nothing_raised {g << "H\txx:i:2" }
    g << "H\tTS:i:120"
    assert_nothing_raised {g << "H\tTS:i:120" }
    assert_raise(RGFA::InconsistencyError) { g << "H\tTS:i:122" }
  end

  def test_rgfa_multiple_def_tags
    g = RGFA.new
    4.times do |i|
      g << "H\txx:i:#{i}".to_rgfa_line
    end
    assert_equal([:xx], g.header.tagnames)
    assert_equal([0,1,2,3], g.header.xx)
    assert_equal([0,1,2,3], g.header.get(:xx))
    assert_equal(:i, g.header.get_datatype(:xx))
    assert_nothing_raised { g.header.validate_field(:xx) }
    [0,2,3].each {|i| g.header.xx.delete(i)}
    g.header.xx = (g.header.xx += [4])
    assert_raise(RGFA::TypeError) { g.header.validate_field(:xx) }
    g.header.xx = g.header.xx.to_rgfa_field_array(:i)
    assert_nothing_raised { g.header.validate_field(:xx) }
    assert_equal([1,4], g.header.get(:xx))
    assert_equal("1\t4", g.header.field_to_s(:xx))
    assert_equal("xx:i:1\txx:i:4", g.header.field_to_s(:xx, tag: true))
    assert_equal(["H\txx:i:1","H\txx:i:4"], g.headers.map(&:to_s))
    g.header.add(:xx, 12)
    g.header.add(:yy, 13)
    assert_equal([1,4,12], g.header.xx)
    assert_equal(13, g.header.yy)
  end

  def test_rgfa_single_def_tags
    g = RGFA.new
    g << "H\txx:i:1".to_rgfa_line
    assert_equal([:xx], g.header.tagnames)
    assert_equal(1, g.header.xx)
    g.header.set(:xx, 12)
    assert_equal(12, g.header.xx)
    g.header.delete(:xx)
    assert_equal(nil, g.header.xx)
  end

end

File: tests/test_api_header.py
import unittest
import gfapy

class TestApiHeader(unittest.TestCase):

  def test_gfa_header(self):
    g = gfapy.Gfa()
    assert(isinstance(g.header, gfapy.line.Header))
    self.assertEqual([], g.header.tagnames)
    g.add_line("H\txx:i:1")
    self.assertEqual(["xx"], g.header.tagnames)

  def test_gfa_header_line_connect(self):
    g = gfapy.Gfa()
    line = gfapy.Line("H\txx:i:1")
    self.assertRaises(gfapy.RuntimeError, line.connect, g)
    g.add_line(line) # nothing raised

  def test_header_version_editing(self):
    standalone = gfapy.Line("H\txx:i:1\tVN:Z:1.0")
    standalone.VN = "2.0" # nothing raised
    g = gfapy.Gfa()
    g.add_line("H\txx:i:1\tVN:Z:1.0")
    g.header.xx = 2 # nothing raised
    with self.assertRaises(gfapy.RuntimeError):
      g.header.VN = "2.0"

  def test_error_inconsistent_definitions(self):
    g = gfapy.Gfa()
    g.add_line("H\txx:i:1")
    g.add_line("H\txx:i:2") # nothing raised
    g.add_line("H\tTS:i:120")
    g.add_line("H\tTS:i:120") # nothing raised
    self.assertRaises(gfapy.InconsistencyError, g.add_line, "H\tTS:i:122")

  def test_gfa_multiple_def_tags(self):
    g = gfapy.Gfa()
    for i in range(4):
      g.add_line("H\txx:i:{}".format(i))
    self.assertEqual(["xx"], g.header.tagnames)
    self.assertEqual([0,1,2,3], g.header.xx)
    self.assertEqual([0,1,2,3], g.header.get("xx"))
    self.assertEqual("i", g.header.get_datatype("xx"))
    g.header.validate_field("xx") # nothing raised
    for i in [0,2,3]:
      g.header.xx.remove(i)
    g.header.xx = [1, 4]
    self.assertRaises(gfapy.TypeError, g.header.validate_field, "xx")
    g.header.xx = gfapy.FieldArray("i", data = g.header.xx)
    g.header.validate_field("xx") # nothing raised
    self.assertEqual([1,4], g.header.get("xx"))
    self.assertEqual("1\t4", g.header.field_to_s("xx"))
    self.assertEqual("xx:i:1\txx:i:4", g.header.field_to_s("xx", tag=True))
    self.assertEqual(sorted(["H\txx:i:1","H\txx:i:4"]),
                     sorted([str(h) for h in g.headers]))
    g.header.add("xx", 12)
    g.header.add("yy", 13)
    self.assertEqual([1,4,12], g.header.xx)
    self.assertEqual(13, g.header.yy)

  def test_gfa_single_def_tags(self):
    g = gfapy.Gfa()
    g.add_line("H\txx:i:1")
    self.assertEqual(["xx"], g.header.tagnames)
    self.assertEqual(1, g.header.xx)
    g.header.set("xx", 12)
    self.assertEqual(12, g.header.xx)
    g.header.delete("xx")
    self.assertEqual(None, g.header.xx)
